Implement initial MailMover test suite

kyrias opened the pull request:
I believe that this should cover all possible rule cases and `max_age`.

This is in preparation for another attempt on #147.

kyrias commented:
Hmm, it only fails on Python 3.4 now since it didn't have `run`, 3.4 went end-of-life in March, so does it need to be supported by afew?

flokli commented:
@kyrias I think only keeping around Python 3.6 and 3.7 are fine.

I openend https://github.com/afewmail/afew/pull/245 for that.

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,152 @@
+
+class TestMailMover(unittest.TestCase):
+    def setUp(self):
+        self.test_dir = tempfile.mkdtemp()
```
Using `tempfile` should be enough - they honor `TEMP` if we really want to change the path.

kyrias commented:
I've rebased and changed so that the mailbox comparisons are done using message IDs rather than filenames, since mids are both more stable, and there's no actual guarantee that the `Maildir` class will use the filenames as keys.  Will squash that into the second commit once approved.

kyrias commented:
And now it passed on both 3.6 and 3.7.  :)

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+import unittest
+
+
+def get_notmuch_db(db_dir):
```
isn't that some basic plumbing for which we also have `afew/Database.py`?

kyrias replied:
There was some reason for not using it originally, but I think I already removed that reason, so fixed.

flokli replied:
:+1:

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+
+def create_mail(msg, maildir, notmuch_db, tags, old=False):
+    if old:
+        date = format_datetime(localtime() - datetime.timedelta(days=20))
```
For the sake of reproducibility, can we hardcode some random date in here, instead of using the current time?

kyrias replied:
We could, but what would we then pass to `MailMover` as the `max_age` argument?  The only real option would be to then also add the functionality to tell MailMover a fixed date as the `max_age` value.

I'm not sure how this would affect reproducibility though?

flokli replied:
To diff multiple test run outputs.

Can we use `localtime()` in the test, but use `@freeze_time` from [freezegun](https://pypi.org/project/freezegun/)?

kyrias replied:
Hm, interesting!  I'll try that out.

kyrias replied:
@flokli There are some issues with this due to the fact that freezegun apparently doesn't handle timezones well.

With the following decorator:

```
@freeze_time("2019-01-30 12:00:00", auto_tick_seconds=15, tz_offset=0)
```

```
now = datetime.now()                 # 2019-01-30 12:00:00
now.strftime('%s')                   # 1548846000
```

And the problem here is that `1548846000` is 12:00 in UTC+1, which is 11:00 UTC.

Meanwhile the date in the mail ends up as `Date: Wed, 30 Jan 2019 13:01:45 +0100` which correctly is 12:00 UTC in UTC+1.  (If switching to `datetime.now()` instead of `localtime()` you get `Date: Wed, 30 Jan 2019 12:01:45 -0000`.)

This means that the date email is an hour ahead of the time returned by `datetime.now()`, which means that it's not included in the timestamp range passed to notmuch.

kyrias replied:
I guess the solution would be to set a static date for the mails created, and it won't be a problem.

kyrias replied:
@flokli 0b37e21 is my attempt at this now.  Is it sufficiently understandable?

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+    else:
+        date = format_datetime(localtime())
+    msgid = make_msgid()
+    email_message = email.message_from_string("""\
```
Can we use `email.message.EmailMessage` to construct the email, instead of assembling the string ourselves?

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+
+    fname = os.path.join(maildir._path, maildir._lookup(message_key))
+    # notmuch 0.26 added index_file and deprecated add_message.
+    if hasattr(notmuch.Database, 'index_file'):
```
Maybe we can use `afew.Database.add_message`, and add the code there?

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+    for tag in tags:
+        notmuch_msg.add_tag(tag, False)
+
+    return (msgid[1:-1], msg)
```
what are we doing here, and why?

kyrias replied:
Stripping the angle brackets off of the message ID that `make_msgid` generated, since the notmuch `get_message_id` function doesn't include them.  I switched to using `strip('<>')` for this instead though.

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+        os.environ['MAILDIR'] = self.test_dir
+
+        self.notmuch_config = os.path.join(self.test_dir, 'notmuch-config')
+        with open(self.notmuch_config, 'w+') as fp:
```
There's `afew/NotmuchSettings.py` - Maybe we can add this code there?

kyrias replied:
Hm, you mean like some way to pass it a dict and then have that written as the config?  Could do that, though I'm not sure if it would be used by anything other than this one place in the code.

kyrias replied:
Added `write_notmuch_settings`, and just set `notmuch_settings` instead of doing it in a more roundabout way.

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+        os.environ['NOTMUCH_CONFIG'] = self.notmuch_config
+
+        # Create notmuch database
+        notmuch.Database(self.test_dir, create=True).close()
```
use `afew.Database`.

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+        self.archive = self.root.add_folder('archive')
+        self.spam = self.root.add_folder('spam')
+
+        self.rules = {
```
It's not really obvious what this structure does and how we assert on it in the tests below. Can you add some comment and elaborate a bit?

kyrias replied:
How's that comment?

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,154 @@
+
+
+    def tearDown(self):
+        shutil.rmtree(self.test_dir)
```
If we're doing everything inside a temporary directory, we can probably use the tempdir contextmanager to let it cleanup for us.

kyrias replied:
That would require each test doing all of the setup as part of the test function instead of using the `unittest` framework's `setUp` function though, but I can do that if you wish I guess?

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,162 @@
+
+    def assert_on_expected(self):
+        with Database() as db:
+            in_inbox = { (os.path.basename(msg.get_message_id()), msg.get_part(1).decode())
```
Maybe this can even be a staticmethod not acessing `self` at all, and we remove `assert_on_expected` entirely.

kyrias replied:
Added a staticmethod to get the folder content and instead have the `assertEqual` calls within the actual test functions now.

flokli replied:
Thanks. This is so much more readable now :)

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,162 @@
+        shutil.rmtree(self.test_dir)
+
+
+    def assert_on_expected(self):
```
I don't really like the way tests mutate `self.expect_*`, and we call that function afterwards. See for a alternative suggestion below.

kyrias commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,162 @@
+        self.notmuch_config = os.path.join(self.test_dir, 'notmuch-config')
+        with open(self.notmuch_config, 'w+') as fp:
+            fp.write('[database]\npath={}\n\n[new]\ntags=new'.format(self.test_dir))
+        os.environ['NOTMUCH_CONFIG'] = self.notmuch_config
```
Yes, it being required for `MailMover` is why it needs to be set as an environment variable.

flokli replied:
And it is required due to `MailMover` shelling out to `notmuch` - I don't see anything else using `NOTMUCH_CONFIG` there. Ok :-)

flokli commented on afew/tests/test_mailmover.py:
```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+# SPDX-License-Identifier: ISC
+# Copyright (c) Johannes Löthberg <johannes@kyriasis.com>
```
Are you fine with dropping the copyright line, for the same reasons as in https://github.com/systemd/systemd/issues/9320:
> Information about who contributed to which file is already covered by `git log` and` git blame` much better. The "copyright by" lines that added manually are often out of date and just unnecessary boilerplate. In particular, all works are covered by copyright without any explicit notice since the 1971 Berne Convention.


I also plan to remove the `# -*- coding: utf-8 -*-` line, as it's a default in Python 3 anyways, so only the `SPDX-License-Identifier: ISC` needs to be preserved.

flokli replied:
`# -*- coding: utf-8 -*-` dropped in #247.

kyrias replied:
Done.
